src/Microsoft.AspNetCore.Server.Kestrel/Internal/Http/Frame.cs

```diff
@@ -598,6 +604,34 @@ public Task WriteAsync(ArraySegment<byte> data, CancellationToken cancellationTo
             }
         }
 
+        private void VerifyWrite()
+        {
+            TryInitializeResponseContentLength();
+
+            if (_responseContentLength.HasValue && _responseBytesWritten > _responseContentLength.Value)
+            {
+                ThrowContentLengthMismatchException();
+            }
+        }
+
+        protected void TryInitializeResponseContentLength()
+        {
+            if (!_responseContentLength.HasValue)
+            {
+                var responseHeaders = FrameResponseHeaders;
+                if (responseHeaders != null && responseHeaders.HasContentLength)
+                {
+                    _responseContentLength = int.Parse(responseHeaders.HeaderContentLength.ToString());
```

Comment:
In that case you should throw from the header setter ideally.

I think we should make sure we have the [same logic we use in HttpAbstractions](https://github.com/aspnet/HttpAbstractions/blob/45696535048559188dce8b38a9fde82baa88893c/src/Microsoft.AspNetCore.Http/Internal/ParsingHelpers.cs#L407).  We should probably look into sharing this with HttpAbsractions somehow.

It's also a good opportunity to make sure we use the same logic for parsing the request header content-length. Right now, it seems like we're probably more lax than we should be.